subsys/net/lib/zperf/zperf_common.c

```diff
@@ -0,0 +1,127 @@
+ */
+
+#include <zephyr/logging/log.h>
+#include <zephyr/net/socket.h>
+
+#include "zperf_internal.h"
+#include "zperf_session.h"
+
+LOG_MODULE_DECLARE(net_zperf, CONFIG_NET_ZPERF_LOG_LEVEL);
+
+#define ZPERF_WORK_Q_THREAD_PRIORITY K_LOWEST_APPLICATION_THREAD_PRIO
+#define ZPERF_WORK_Q_STACK_SIZE 2048
+
+K_THREAD_STACK_DEFINE(zperf_work_q_stack, ZPERF_WORK_Q_STACK_SIZE);
+
+static struct k_work_q zperf_work_q;
+
+int zperf_prepare_upload_sock(const struct sockaddr *peer_addr, int tos,
+			      int proto)
+{
+	socklen_t addrlen = peer_addr->sa_family == AF_INET6 ?
+			    sizeof(struct sockaddr_in6) :
+			    sizeof(struct sockaddr_in);
+	int type = (proto == IPPROTO_UDP) ? SOCK_DGRAM : SOCK_STREAM;
+	int sock = -1;
+	int ret;
+
+	switch (peer_addr->sa_family) {
+	case AF_INET:
+		if (!IS_ENABLED(CONFIG_NET_IPV4)) {
+			NET_ERR("IPv4 not available.");
+			return -EINVAL;
+		}
+
+		sock = zsock_socket(AF_INET, type, proto);
+		if (sock < 0) {
+			NET_ERR("Cannot create IPv4 network socket (%d)",
+				errno);
+			return -errno;
+		}
+
+		if (tos > 0) {
+			if (zsock_setsockopt(sock, IPPROTO_IP, IP_TOS,
+					     &tos, sizeof(tos)) != 0) {
+				NET_WARN("Failed to set IP_TOS socket option. "
+					 "Please enable CONFIG_NET_CONTEXT_DSCP_ECN.");
+			}
+		}
+
+		break;
+
+	case AF_INET6:
+		if (!IS_ENABLED(CONFIG_NET_IPV6)) {
+			NET_ERR("IPv6 not available.");
+			return -EINVAL;
+		}
+
+		sock = zsock_socket(AF_INET6, type, proto);
+		if (sock < 0) {
+			NET_ERR("Cannot create IPv6 network socket (%d)",
+				errno);
+			return -errno;
+		}
+
+		if (tos >= 0) {
+			if (zsock_setsockopt(sock, IPPROTO_IPV6, IPV6_TCLASS,
+					     &tos, sizeof(tos)) != 0) {
+				NET_WARN("Failed to set IPV6_TCLASS socket option. "
+					 "Please enable CONFIG_NET_CONTEXT_DSCP_ECN.");
+			}
+		}
+
+		break;
+
+	default:
+		LOG_ERR("Invalid address family (%d)", peer_addr->sa_family);
+		return -EINVAL;
+	}
+
+	ret = zsock_connect(sock, peer_addr, addrlen);
+	if (ret < 0) {
+		NET_ERR("Connect failed (%d)", errno);
+		ret = -errno;
+		zsock_close(sock);
+		return ret;
+	}
+
+	return sock;
+}
+
+uint32_t zperf_packet_duration(uint32_t packet_size, uint32_t rate_in_kbps)
+{
+	return (uint32_t)(((uint64_t)packet_size * 8U * USEC_PER_SEC) /
+			  (rate_in_kbps * 1024U));
+}
+
+void zperf_async_work_submit(struct k_work *work)
+{
+	k_work_submit_to_queue(&zperf_work_q, work);
+}
+
+static int zperf_init(const struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	k_work_queue_init(&zperf_work_q);
+	k_work_queue_start(&zperf_work_q, zperf_work_q_stack,
+			   K_THREAD_STACK_SIZEOF(zperf_work_q_stack),
+			   ZPERF_WORK_Q_THREAD_PRIORITY, NULL);
+	k_thread_name_set(&zperf_work_q.thread, "zperf_work_q");
+
+	zperf_udp_uploader_init();
+	zperf_tcp_uploader_init();
+	zperf_udp_receiver_init();
+	zperf_tcp_receiver_init();
+
+	zperf_session_init();
+	zperf_shell_init();
+
+	return 0;
+}
+
+SYS_INIT(zperf_init, APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```

Comment:
On boards where the network interface gets initialized later in the boot sequence, leaving that in outputs the following errors:
```
[00:00:00.004,000] <wrn> net_if: iface 0x200013c8 is down
[00:00:00.004,000] <err> net_if: Cannot join all nodes address ff02::1 (-1)
[00:00:00.004,000] <wrn> net_if: iface 0x200013c8 is down
[00:00:00.004,000] <err> net_if: Cannot join solicit node address ff02::1:ff00:1 (-1)
```

The async api still works but that error can be misleading. Is there a way to delay the zperf initialization to after the interface has been brought up? I am using the STM32G474re with ENC424J600.

Reply:
In this case the init could be placed to `net_core.c` `services_init()` function.

Reply:
It's just a notification that the stack failed to send MLD notification nothing critical. Probably it'd be best to rework how those notfications are sent, so that the stack would not even attempt to do so if the interface is down (similarily how DAD is handled - if the interface is down, DAD will be performed after the interface is brought up).

> In this case the init could be placed to net_core.c services_init() function.

I'm afraid this won't help as `net_core` is initialized even earlier than `zperf` (POST_KERNEL vs APPLICATION).

Anyway, after looking more closely into the code, I wonder if it's even needed for `zperf` to register IP addresses at boot. It'd make sense to do that for `set_ip` command, but at boot, it's either address provided by NET_CONFIG, which is added to an interface anyway, or nothing. I think I'll rework the initialization to skip that part.

Reply:
Thank you the warning is gone!